### ISSUE TITLE
Integration testing examples and documentation

### ISSUE THREAD

**Issue by WillSquire** [diesel-rs/diesel#2058]
I've been reading through the examples and docs but I can't find material on integration testing? Found a bit of information [here](https://users.rust-lang.org/t/unit-testing-diesel-code/21196) regarding `Connection::test_transaction`. Is there material on this?

This is my code that contains the database connection pool:
```rust
use crate::error::Error;
use diesel::{pg::PgConnection, r2d2::ConnectionManager};
use r2d2::{Pool, PooledConnection};

pub type Connection = PooledConnection<ConnectionManager<PgConnection>>;
type ConnectionPool = Pool<ConnectionManager<PgConnection>>;

pub struct Db {
  connection_pool: ConnectionPool,
}

impl Db {
  pub fn new(
    db_user: &str,
    db_password: &str,
    db_name: &str,
    db_server: &str,
  ) -> Result<Db, Error> {
    Ok(Db {
      connection_pool: Pool::builder().max_size(15).build(
        ConnectionManager::<PgConnection>::new(format!(
          "postgres://{}:{}@{}/{}",
          db_user, db_password, db_server, db_name
        )),
      )?,
    })
  }

  pub fn connect(&self) -> Result<Connection, Error> {
    Ok(self.connection_pool.clone().get()?)
  }
}
```

For brevity the actual usage can be found [here](https://github.com/WillSquire/Graphy/blob/9299a7e7933304b1602a2bb401a61d20f06e01ba/api/src/models/user.rs), which I might get away with passing a test connection if it exists.

Noticed the r2d2 feature uses `TestConnection` from `test_helpers::*`. Is the the recommended way?

**Comment by weiznich** [comment]
The following code will call `begin_test_transaction` for each connection in a given pool:

```rust
struct TestTransaction;

impl CustomizeConnection<DbConnection, ::diesel::r2d2::Error> for TestTransaction {
    fn on_acquire(
        &self,
        conn: &mut DbConnection,
    ) -> ::std::result::Result<(), ::diesel::r2d2::Error> {
        conn.begin_test_transaction().unwrap();
        Ok(())
    }
}

let manager = ConnectionManager::<DbConnection>::new(db_url);
    let pool = Pool::builder()
        .connection_customizer(Box::new(TestTransaction))
        .build(manager)
        .expect("Failed to init pool")
```

**Comment by WillSquire** [comment]
Thanks @weiznich, I've tried implementing what you've put above but I'm getting `no method named 'begin_test_transaction' found for type '&mut diesel::PgConnection' in the current scope`. Is this implemented for postgres or am I using the wrong connection type?

**Comment by weiznich** [comment]
Just to add the solution also to this issue:
You need to import the `Connection` trait into the current scope, otherwise `begin_test_transaction` is not found.

**Comment by Mastermindaxe** [comment]
I think a guide on testing would be very helpful. Is there any work done on this since posting this issue? I guess this is also related to issue #1549 as it seems there is documentation on testing overall missing. Thanks in advance

**Comment by WillSquire** [comment]
@Mastermindaxe I'm not sure about documentation, but I did implement it and have an example of the path I took here: https://github.com/WillSquire/Graphy/blob/master/api/src/db/mod.rs . Not worked on it for perhaps a year though.

**Comment by weiznich** [comment]
Closed as this seems to be a duplicate of #777
